src/engine/QueryPlanner.cpp

```diff
@@ -1693,6 +1713,47 @@ std::vector<QueryPlanner::SubtreePlan> QueryPlanner::createJoinCandidates(
   return candidates;
 }
 
+// _____________________________________________________________________________
+auto QueryPlanner::createSpatialJoin(
+    const SubtreePlan& a, const SubtreePlan& b,
+    const std::vector<std::array<ColumnIndex, 2>>& jcs)
+    -> std::optional<SubtreePlan> {
+  auto aIsSpatialJoin =
+      std::dynamic_pointer_cast<const SpatialJoin>(a._qet->getRootOperation());
+  auto bIsSpatialJoin =
+      std::dynamic_pointer_cast<const SpatialJoin>(b._qet->getRootOperation());
+
+  bool aIs = static_cast<bool>(aIsSpatialJoin);
+  bool bIs = static_cast<bool>(bIsSpatialJoin);
+
+  if ((aIs && bIs) || (!aIs && !bIs)) {
+    return std::nullopt;
+  }
+
+  SubtreePlan spatialSubtreePlan = aIsSpatialJoin ? a : b;
+  SubtreePlan otherSubtreePlan = aIsSpatialJoin ? b : a;
+
+  std::shared_ptr<Operation> op = spatialSubtreePlan._qet->getRootOperation();
+  SpatialJoin* spatialJoin = static_cast<SpatialJoin*>(op.get());
+
+  if (spatialJoin->isConstructed()) {
+    return std::nullopt;
+  }
+
+  if (jcs.size() > 1) {
+    AD_THROW("in its current implementation only one pair is allowed");
+  }
+  ColumnIndex ind = aIsSpatialJoin ? jcs[0][1] : jcs[0][0];
+  Variable var =
+      otherSubtreePlan._qet->getVariableAndInfoByColumnIndex(ind).first;
+
+  auto newSpatialJoin = spatialJoin->addChild(otherSubtreePlan._qet, var);
+
+  SubtreePlan plan = makeSubtreePlan<SpatialJoin>(newSpatialJoin);
+  mergeSubtreePlanIds(plan, a, b);
+  return plan;
+}
+
```

Comment:
You should also write tests for the query planning, see the QueryPlannerTest.cpp and QueryPlannerTestHelpers.cpp for tons of examples how to do this elegantly.